dev/com.ibm.ws.logging.core/src/com/ibm/ws/collector/manager/buffer/BufferManagerImpl.java

```diff
@@ -16,278 +16,278 @@
-                if (lastReportTimeForWarning == 0) {
-                    // This was the first instance where we started to see loss
-                    // of events.
-                    Tr.warning(tc, "HANDLER_STARTED_TO_LOSE_EVENTS_WARNING", handlerId, lostEventsForWarning, sourceId);
-                } else {
-                    long timeElapsedInMins = TimeUnit.MILLISECONDS.toMinutes(timeElapsed);
-                    Tr.warning(tc, "HANDLER_LOST_EVENTS_WARNING", handlerId, lostEventsForWarning, sourceId,
-                            timeElapsedInMins, totalLostEvents);
-                }
-                lastReportTimeForWarning = currentTime;
-                lostEventsForWarning = 0;
-            }
-            timeElapsed = currentTime - lastReportTimeForTrace;
-            if (timeElapsed >= intervalForTrace) {
-                if (TraceComponent.isAnyTracingEnabled() && tc.isEventEnabled()) {
-                    long timeElapsedInSecs = TimeUnit.MILLISECONDS.toSeconds(timeElapsed);
-                    Tr.event(tc,
-                            "Handler [{0}] has lost {1} events from source [{2}] in the last {3} second(s), and has lost {4} events from the source since the handler started.",
-                            handlerId, lostEventsForTrace, sourceId, timeElapsedInSecs, totalLostEvents);
-                }
-                lastReportTimeForTrace = currentTime;
-                lostEventsForTrace = 0;
-            }
-        }
-
-    }
+	private static final TraceComponent tc = Tr.register(BufferManagerImpl.class);
+	private Buffer<Object> ringBuffer;
+	private Set<SynchronousHandler> synchronousHandlerSet = new HashSet<SynchronousHandler>();
+
+	private final int capacity;
+
+	private final String sourceId;
+	/* Map to keep track of the next event for a handler */
+	private final ConcurrentHashMap<String, HandlerStats> handlerEventMap = new ConcurrentHashMap<String, HandlerStats>();
+
+	protected Queue<Object> earlyMessageQueue;
+
+	private static final int EARLY_MESSAGE_QUEUE_SIZE = 400;
+
+	public BufferManagerImpl(int capacity, String sourceId) {
+		super();
+
+		BufferManagerEMQHelper.addBufferManagerList(this);
+		ringBuffer = null;
+		this.sourceId = sourceId;
+		this.capacity = capacity;
+		if (!BufferManagerEMQHelper.getEMQRemovedFlag()) {
+			earlyMessageQueue = new SimpleRotatingSoftQueue<Object>(new Object[EARLY_MESSAGE_QUEUE_SIZE]);
+			//Check again just in case
+			if (BufferManagerEMQHelper.getEMQRemovedFlag()){
+				removeEMQ();
+			}
+		}
+	}
+
+	public BufferManagerImpl(int capacity, String sourceId, boolean isEMQ) {
+		super();
+		this.sourceId = sourceId;
+		this.capacity = capacity;
+		if (!isEMQ) {
+			earlyMessageQueue = null; // don't need earlyMessageQueue
+			ringBuffer = new Buffer<Object>(capacity);
+		}
+	}
+
+	@Override
+	public void add(Object event) {
+		if (event == null)
+			throw new NullPointerException();
+
+		
+		if (ringBuffer != null) {
```

Comment:
what if ringBuffer gets added between lines 74 and 89?  Seems like the new async handler would miss out on this event

to fix that you might need to put synchronized block around lines 74-81 and get rid of the "NPE approach" to adding to ringBuffer

also, if ringBuffer is used outside of synchronized block you might need to make it volatile